### ISSUE TITLE
My implementation of NeighborSampling is unfortunately slow

### ISSUE THREAD

**Issue by dcoukos** [pyg-team/pytorch_geometric#1034]
## ❓ Questions & Help

Recently I have been trying to implement FeaStNet-style models to graphs which represent molecule surfaces. The problem is that in order to capture relevant information from the surface, I would like the model to be able to see at least 5 hops from each node.

After implementing a non-bipartite version of the model I found performance to be very slow. I am likely doing something very wrong, because even when I implement this with just two hops, I get a relative slowdown of 1500x with respect to TwoHop. 

Data comprises about 2000 graphs, each with about 6000 nodes.

An example of the training loop:

    for batch in tqdm(train_loader):
        ns = NeighborSampler(batch, size=[0.5], num_hops=5, bipartite=False)

        for subdata in ns():
            optimizer.zero_grad()
            out = model(batch.x[subdata.n_id], subdata.edge_index)
            label = batch.y[subdata.n_id].to(device).view(-1, 1)
            tr_loss = F.binary_cross_entropy_with_logits(out, target=label)
            tr_loss.backward()
            optimizer.step()

I have tried to base my code off of https://github.com/rusty1s/pytorch_geometric/blob/a8f32aaff8608e497f112f700d1fd8ca0cb9ae18/test/data/test_sampler.py#L96
Is there anything obviously wrong above?
Thanks for any help with resolving this!

**Comment by rusty1s** [comment]
Hi, it might be best to discuss this in a single issue. My experience is that neighborhood sampling simply isn't useful when applied to many hops. Memory consumptions are huge to just train on a single node, and you need specialized architectures where something like skip connections just feel hacky to implement. In my experience, there are simply better ways to get the job done. For example, since your graphs have only about 6000 nodes, I do not see why you even need to make use of node sampling. If you really want to make use of node sampling, alternatives such as Cluster-GCN (just recently added to PyG) scale way better.

**Comment by dcoukos** [comment]
Hi, 
Thanks for the quick reply!
I may have misunderstood the concept of neighborhood sampling...
So the reason I was interested in node sampling is that protein molecular surfaces are organized in such a way that relevant features for the classification of my node can be stored in nodes multiple hops away. I don't think using a coarser graph would not work since important surface shapes may be lost. Additionally, implementing deep networks does not seem to work, possible because the information is getting "smoothed out" as it goes through successive convolutional layers. If I were to imagine a similar set up on images, I would want to use filters that have a larger view (perhaps with dilation), since it might be necessary to directly compare features situated several pixels apart. 

As I understood it, node sampling essentially leads to the lazy generation of message passing networks that are functionally equivalent to subregions of the graph but are highly connected up to n hops. Is this incorrect?

Thanks for mentioning Cluster-GCN, I'll go take a look at it now!

Also, I would like to apologize for my description of the slow down, a large part of the problem was the batch_size, which I had left at the default of 1. If I increase this to 10,000 or 100,000 to implement two hops (just for comparison's sake) I get much better performance. A 20x slowdown with respect to simply using a TwoHop transform persists, but I understand this isn't the right use case.
